File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def dfs(numbers, index, next_num, target):
    count = 0
    if len(numbers) == index:
        if target == next_num:
            return 1
        return 0
    count += dfs(numbers, index+1, next_num + numbers[index], target)
    count += dfs(numbers, index+1, next_num - numbers[index], target)
    return count
    
    
def solution(numbers, target):
    return dfs(numbers, 0, 0, target)

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;

class Solution {
    private int dfs(int current, int target, int[] numbers, int i){
        if(i == numbers.length){
            // 모든 수 다 쓰고 current가 target인지 확인하기
            return current == target ? 1 : 0;
        }
        
        int plus = dfs(current + numbers[i], target, numbers, i+1);
        int minus = dfs(current - numbers[i], target, numbers, i+1);
        return plus + minus;
    }
    
    public int solution(int[] numbers, int target) {
        return dfs(0, target, numbers, 0);
    }
}